tuxedo-template-runtime/src/kitties.rs

```diff
@@ -1,126 +1,544 @@
+        child: &KittyData,
+    ) -> Result<(), Self::Error>;
+}
+
+pub struct KittyHelpers;
+impl Breed for KittyHelpers {
+    const COST: u128 = 5u128;
+    const NUM_FREE_BREEDINGS: u64 = 2u64;
+    type Error = VerifierError;
+    /// Checks:
+    ///     - Mom can breed
+    ///     - Dad can breed
+    ///
+    fn can_breed(mom: &KittyData, dad: &KittyData) -> Result<(), Self::Error> {
+        Self::check_mom_can_breed(mom)?;
+        Self::check_dad_can_breed(dad)?;
+        Self::check_free_breedings(mom, dad)?;
         Ok(())
     }
-}
 
-impl UtxoSet for KittiesPiece {
-    /// Check whether a given utxo exists in the current set
-    fn contains(utxo_ref: UtxoRef) -> bool {
-        sp_io::storage::exists(&utxo_ref.encode())
+    /// Checks:
+    ///     - Mom is in `RearinToGo` state
+    ///     - Mom number of breedings is not maxed out
+    ///
+    fn check_mom_can_breed(mom: &KittyData) -> Result<(), Self::Error> {
+        match &mom.parent {
+            Parent::Mom(status) => {
+                if let MomKittyStatus::HadBirthRecently = status {
+                    return Err(Self::Error::MomNotReadyYet);
+                }
+            }
+            Parent::Dad(_) => return Err(Self::Error::TwoDadsNotValid),
+        }
+        mom.num_breedings
+            .checked_add(1)
+            .ok_or(Self::Error::TooManyBreedingsForKitty)?;
+        Ok(())
     }
 
-    /// Insert the given utxo into the state storing it with the given ref
-    /// The ref is probably the hash of a tx that created it and its index in that tx, but this decision is opaque to this trait
-    /// Return whether the operation is successful (It can fail if the ref is already present)
-    fn insert(utxo_ref: UtxoRef, utxo: &Utxo) -> bool {
-        sp_io::storage::set(&utxo_ref.encode(), &utxo.encode());
-        true
+    /// Checks:
+    ///     - Dad is in `RearinToGo` state
+    ///     - Dad number of breedings is not maxed out
+    ///
+    fn check_dad_can_breed(dad: &KittyData) -> Result<(), Self::Error> {
+        match &dad.parent {
+            Parent::Dad(status) => {
+                if let DadKittyStatus::Tired = status {
+                    return Err(Self::Error::DadTooTired);
+                }
+            }
+            Parent::Mom(_) => return Err(Self::Error::TwoMomsNotValid),
+        }
+        dad.num_breedings
+            .checked_add(1)
+            .ok_or(Self::Error::TooManyBreedingsForKitty)?;
+        Ok(())
     }
 
+    /// Checks:
+    ///     - Both parents free breedings is non-zero
     ///
-    /// nullify the utxo by either:
-    /// - Consuming it entirely
-    /// - Putting it on "Timeout"
-    /// - Not consuming it but marking it as spent
+    fn check_free_breedings(mom: &KittyData, dad: &KittyData) -> Result<(), Self::Error> {
+        let mom_breedings = mom.free_breedings;
+        let dad_breedings = dad.free_breedings;
+        if (mom_breedings == 0) && (dad_breedings == 0) {
```

Comment:
Should this be `||`? Doesn't it count as a breeding for each of them?

Reply:
True great catch hopefully tests would have gotten it